### ISSUE TITLE
HUGETLB_FORCE_ELFMAP appears broken

### ISSUE THREAD

**Issue by anadav** [libhugetlbfs/libhugetlbfs#96]
tl;dr: I am trying to use `HUGETLB_FORCE_ELFMAP.` It does not appear to work properly, and my debugging hints there are a couple of issues. When I fix the issues (below), things seem to work correctly. However, my confidence is slightly low since the faulty code (IIUC) seems to be in since day 1.

I have allocated huge-pages and run: `HUGETLB_ELFMAP=R HUGETLB_FORCE_ELFMAP=yes  HUGETLB_VERBOSE=3 LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libhugetlbfs.so binary`

I get:

```libhugetlbfs: INFO: Found pagesize 2048 kB
libhugetlbfs: INFO: Parsed kernel version: [6] . [8] . [0] 
libhugetlbfs: INFO: Feature private_reservations is present in this kernel
libhugetlbfs: INFO: Feature noreserve_safe is present in this kernel
libhugetlbfs: INFO: Feature map_hugetlb is present in this kernel
libhugetlbfs: INFO: Kernel has MAP_PRIVATE reservations.  Disabling heap prefaulting.
libhugetlbfs: INFO: Kernel supports MAP_HUGETLB
libhugetlbfs: INFO: HUGETLB_FORCE_ELFMAP=yes, enabling partial segment remapping for non-relinked binaries
libhugetlbfs: INFO: Disabling filesz copy optimization
libhugetlbfs: INFO: HUGETLB_SHARE=0, sharing disabled
libhugetlbfs: INFO: HUGETLB_NO_RESERVE=no, reservations enabled
libhugetlbfs: INFO: Segment 2's aligned memsz is too small: 0x25588 < 0xffffffffffffffff
libhugetlbfs: INFO: Segment 3's aligned memsz is too small: 0x469b81 < 0xffffffffffffffff
libhugetlbfs: INFO: Segment 4's aligned memsz is too small: 0xe17d8 < 0xffffffffffffffff
libhugetlbfs: INFO: Segment 5's unaligned memsz is too small: 0x6ecc8 < 0x11d350
libhugetlbfs: INFO: No segments were appropriate for remapping
libhugetlbfs: INFO: Not setting up morecore because it's not available (see issue #52).
```

The failure seems wrong, as the binary's text section is apparently large enough for HUGETLB_FORCE_ELFMAP to work and there are available huge-pages, which are not used:

```
AnonHugePages:         0 kB
ShmemHugePages:        0 kB
FileHugePages:     14336 kB
HugePages_Total:    5000
HugePages_Free:     4998
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:        10240000 kB
```

After some debugging, I think there are two issues in `parse_elf_partial`:

1. slice_end is miscalculated
2. htlb_seg_table[htlb_num_segs].page_size is not initialized

My hacky fix which seems to work is therefore:
```diff --git a/elflink.c b/elflink.c
index 78b0edd..1fb3f0b 100644
--- a/elflink.c
+++ b/elflink.c
@@ -785,7 +785,7 @@ int parse_elf_partial(struct dl_phdr_info *info, size_t size, void *data)
                vaddr = hugetlb_next_slice_start(info->dlpi_addr +
                                                 info->dlpi_phdr[i].p_vaddr);
                gap = vaddr - (info->dlpi_addr + info->dlpi_phdr[i].p_vaddr);
-               slice_end = hugetlb_slice_end(vaddr);
+               slice_end = hugetlb_slice_end(vaddr + 1);
                /*
                 * we should stop remapping just before the slice
                 * containing the end of the memsz portion (taking away
@@ -822,6 +822,7 @@ int parse_elf_partial(struct dl_phdr_info *info, size_t size, void *data)
                htlb_seg_table[htlb_num_segs].vaddr = (void *)vaddr;
                htlb_seg_table[htlb_num_segs].filesz = memsz;
                htlb_seg_table[htlb_num_segs].memsz = memsz;
+               htlb_seg_table[htlb_num_segs].page_size = hpage_readonly_size;
 
                htlb_num_segs++;
        }
```

**Comment by luochenglcs** [comment]
Solved by #101
